### PR TITLE
fix(typeahead): fix focus after leaving control (#4622)

### LINKED ISSUE #1884: fix(typeahead): typeahead focus obtained again after leaving the control
I am experiencing the following issue:

- I setup the control in the following way. Async call is done to an API, which hits an external resource. This operation takes, at least, 1 second. 

  <input class="form-control"
             name="searchForm"
             required
             #searchForm="ngModel"
             [(ngModel)]="searchModel"
             [typeahead]="searchResults"
             (typeaheadOnSelect)="selectFromSearchResult($event)"
             typeaheadOptionField="fullName"
             typeaheadWaitMs="200"
             (keypress)="loading=true"
             size="30" type="text" />

Component code is as follows:

 this.searchResults = Observable.create((observer:any) => {
      this.loading = true;
      this.contactsService.search(this.searchModel)
        .subscribe((result: any) => {
          this.loading = false;
          observer.next(JSON.parse(result._body));
        })
    });

- Upon typing, after 200ms a request is executed.
- I hit tab before the response comes back, so focus is now in another control.
- After the results come back from the server, typeahead control obtains focus again. This is not an expected behaviour.

Q: is there a way to cancel the request? 
Q: shouldn't the control cancel the request on focus lost? 

Thank you in advance!

### PR DESCRIPTION
**Pull request by daniloff200**
Co-authored-by: Dmitriy Danilov <daniloff200@gmail.com>

# PR Checklist
Before creating new PR, please take a look at checklist below to make sure that you've done everything that needs to be done before we can merge it.

 - [x] read and followed the [CONTRIBUTING.md](https://github.com/valor-software/ngx-bootstrap/blob/development/CONTRIBUTING.md) guide.
 - [x] built and tested the changes locally.
 - [ ] added/updated tests.
 - [x] added/updated API documentation.
 - [x] added/updated demos.

That's another PR, which related to #4622, cause, fork was deleted, and it need some fixes
closes #1884

### CODE REVIEW

File: src/typeahead/typeahead.directive.ts [hunk 1]
```diff
@@ -119,6 +119,8 @@ export class TypeaheadDirective implements OnInit, OnDestroy {
 
   /** This attribute indicates that the dropdown should be opened upwards */
   @Input() dropup = false;
+  /** if true, typeahead will cancel async request on blur */
+  @Input() typeaheadCancelOnBlur = false;
```
Comment (valorkin) [inline src/typeahead/typeahead.directive.ts]: this option will not be changed too often, please move to it to config,
to avoid Component bloating

**Review by valorkin — CHANGES_REQUESTED** [review]

File: src/typeahead/typeahead.config.ts [hunk 2]
```diff
@@ -9,6 +9,8 @@ export class TypeaheadConfig {
   isAnimated = false;
   /** used to hide results on blur */
   hideResultsOnBlur = true;
+  /** if true, typeahead will cancel async request on blur */
+  typeaheadCancelOnBlur = false;
```
Comment (valorkin) [inline src/typeahead/typeahead.config.ts]: cancelRequestOnFocusLost

File: src/typeahead/typeahead.directive.ts [hunk 3]
```diff
@@ -168,6 +170,7 @@ export class TypeaheadDirective implements OnInit, OnDestroy {
     Object.assign(this,
       {
         typeaheadHideResultsOnBlur: config.hideResultsOnBlur,
+        typeaheadCancelOnBlur: config.typeaheadCancelOnBlur,
```
Comment (valorkin) [inline src/typeahead/typeahead.directive.ts]: rename here to

**Review by valorkin — CHANGES_REQUESTED** [review]
naming, please

**Review by valorkin — APPROVED** [review]
Looks good to me!

**Comment by dmitry-zhemchugov** [comment]
Tested this PR. Ready for merge.

**Pull request merged**